exporter/otlp/lib/opentelemetry/exporter/otlp/exporter.rb

```diff
@@ -111,7 +110,19 @@ def shutdown(timeout: nil)
 
         private
 
-        def http_connection(uri, ssl_verify_mode, certificate_file)
+        def validate!(args)
+          raise ArgumentError, "invalid url for OTLP::Exporter #{args[:endpoint]}" unless OpenTelemetry::Common::Utilities.valid_url?(args[:endpoint])
+          raise ArgumentError, "unsupported compression key #{args[:compression]}" unless args[:compression].nil? || %w[gzip none].include?(args[:compression])
+          raise ArgumentError, ERROR_MESSAGE_INVALID_HEADERS unless [String, Hash].include?(args[:headers].class)
+        end
+
+        def uri
+          @uri ||= URI.join(@endpoint, 'v1/traces') if @endpoint == ENV['OTEL_EXPORTER_OTLP_ENDPOINT']
```

Comment:
I can't make sense of this condition. Maintainers what do you think?

```ruby
[4] pry(main)> ENV['OTEL_EXPORTER_OTLP_ENDPOINT']='http://localhost:4318/v1/traces'
=> "http://localhost:4318/v1/traces"
[5] pry(main)> OpenTelemetry::Exporter::OTLP::Exporter.new(endpoint: ENV['OTEL_EXPORTER_OTLP_ENDPOINT'])        
=> #<OpenTelemetry::Exporter::OTLP::Exporter:0x000000011a16d328
...
 @path="/v1/v1/traces", # <---- this is obviously a bug

```

Reply:
Apologies, are you asking if you don’t understand the conditional in the code you wrote, or as it was prior to this change?

Reply:
I didn't change the condition, I extracted it from the initializer.

```ruby
          @uri = if endpoint == ENV['OTEL_EXPORTER_OTLP_ENDPOINT']
                   URI.join(endpoint, 'v1/traces')
                 else
                   URI(endpoint)
                 end
```

Reply:
what I don't understand is the idea behind it, why would both `endpoint` and `ENV['OTEL_EXPORTER_OTLP_ENDPOINT']` have to be set to the same value for the URI to have `v1/traces` appended to it?

Reply:
Ah, gotcha. So this is a little subtle - OpenTelemetry defines various environment variables that SDKs must respect, and the ones we're dealing with here are for [the OTLP exporter, specifically](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#configuration-options). Here's the entire blob about endpoint selection:

> Endpoint (OTLP/HTTP): Target URL to which the exporter is going to send spans or metrics. The endpoint MUST be a valid URL with scheme (http or https) and host, MAY contain a port, SHOULD contain a path and MUST NOT contain other parts (such as query string or fragment). A scheme of https indicates a secure connection. When using OTEL_EXPORTER_OTLP_ENDPOINT, exporters MUST construct per-signal URLs as [described below](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#endpoint-urls-for-otlphttp). The per-signal endpoint configuration options take precedence and can be used to override this behavior (the URL is used as-is for them, without any modifications). See the [OTLP Specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp-request) for more details.

> Default: http://localhost:4318 [1]
> Env vars: OTEL_EXPORTER_OTLP_ENDPOINT OTEL_EXPORTER_OTLP_TRACES_ENDPOINT OTEL_EXPORTER_OTLP_METRICS_ENDPOINT

Crucially, our OTLP exporter *only* supports traces right now. So if someone configures `OTEL_EXPORTER_OTLP_ENDPOINT` and gives it `http://my-fancy-exporter.com:4318` as a _generic_ OTLP endpoint, we *must* construct a signal-specific endpoint for traces, which must have the `/v1/traces` bits appended (which is also dictated by the spec).

However, signal-specific endpoints take precedence, so the best way for us to detect this condition is to see if our env-var resolving code (which considers all the relevant env vars) selected the _same_ thing as what is present in `ENV['OTEL_EXPORTER_OTLP_ENDPOINT']`? If so, we assume that the user configured a generic OTLP endpoint and that we must construct the signal-specific endpoint.